sawmill-core/src/main/java/io/logz/sawmill/processors/ExternalMappingSourceProcessor.java

```diff
@@ -0,0 +1,186 @@
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+@ProcessorProvider(type = "externalMapping", factory = ExternalMappingSourceProcessor.Factory.class)
+public class ExternalMappingSourceProcessor implements Processor {
+
+    public static final int MINIMUM_REFRESH_PERIOD_IN_SECONDS = 15;
+    public static final int DISABLE_MAPPING_REFRESH = -1;
+
+    private final Logger logger = LoggerFactory.getLogger(ExternalMappingSourceProcessor.class);
+
+    private final String sourceField;
+    private final String targetField;
+    private final ExternalMappingsClient externalMappingsClient;
+    private final long mappingRefreshPeriodInSeconds;
+
+    private static ScheduledExecutorService scheduledRefreshMappingsExecutor;
+
+    @SuppressWarnings("Guava")
+    private final Supplier<Void> lazyInitSupplier;
+
+    private volatile Map<String, Iterable<String>> keyValueMappingsCache;
+
+    public ExternalMappingSourceProcessor(Configuration configuration) throws MalformedURLException {
+        this.sourceField = requireNonNull(configuration.getSourceField());
+        this.targetField = requireNonNull(configuration.getTargetField());
+        this.externalMappingsClient = new ExternalMappingsClient(configuration);
+        this.mappingRefreshPeriodInSeconds = configuration.getMappingRefreshPeriodInSeconds();
+
+        lazyInitSupplier = Suppliers.memoize(this::lazyInit);
+    }
+
+
+    private Void lazyInit() {
+        refreshExternalMapping();
+        if (mappingRefreshPeriodInSeconds != DISABLE_MAPPING_REFRESH) {
+            initThreadPool();
+            startScheduledExecutor();
+        }
+        /* return value is not used */
+        return null;
+    }
+
+    private void refreshExternalMapping() {
+        try {
+            keyValueMappingsCache = externalMappingsClient.loadMappings();
+        } catch (Exception e) {
+            logger.error("Cannot load external mapping for field {} due to an unexpected error", sourceField, e);
+        }
+
+        if (keyValueMappingsCache == null) {
+            keyValueMappingsCache = new HashMap<>();
+        }
+
+        if (keyValueMappingsCache.isEmpty()) {
+            logger.error("Cannot load external mapping for field: {}, received an empty map", sourceField);
+        }
+    }
+
+    private void initThreadPool() {
+        if (scheduledRefreshMappingsExecutor != null) return;
+        scheduledRefreshMappingsExecutor = Executors.newScheduledThreadPool(1,
+            new ThreadFactoryBuilder().setNameFormat("refresh-mapping-executor").setDaemon(true).build());
+    }
+
+    private void startScheduledExecutor() {
+        scheduledRefreshMappingsExecutor.scheduleAtFixedRate(this::refreshExternalMapping, mappingRefreshPeriodInSeconds,
+            mappingRefreshPeriodInSeconds, TimeUnit.SECONDS);
+        Runtime.getRuntime().addShutdownHook(new Thread(this::shutdownScheduledExecutor, "shutdown-hook-thread"));
```

Comment:
this should be done in `initThreadPool()`, no? this is where you create the executor